src/Illuminate/Queue/Worker.php

```diff
@@ -151,9 +154,20 @@ public function pop($connectionName, $queue = null, $delay = 0, $sleep = 3, $max
         // then immediately return back out. If there is no job on the queue
         // we will "sleep" the worker for the specified number of seconds.
         if (! is_null($job)) {
-            return $this->process(
-                $this->manager->getName($connectionName), $job, $maxTries, $delay
-            );
+            if ($timeout > 0) {
+                pcntl_alarm($timeout);
```

Comment:
@taylorotwell This works by calling `pcntl_alarm($timeout)` and `pcntl_alarm(0)` and this handler:
```php
pcntl_signal(SIGALRM, function () {
    throw new TimeoutException();
}, true);
```
This is [a sample gist](https://gist.github.com/jsjohnst/3297804).